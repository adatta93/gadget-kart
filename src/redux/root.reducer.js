import { combineReducers } from "redux";
import { cartReducer } from "./cart/cart.reducer";
import { fetchProductReducer } from "./products/products.reducer";
import { userReducer } from "./users/users.reducer";

export default combineReducers({
  products: fetchProductReducer,
  user: userReducer,
  cart: cartReducer,
});
