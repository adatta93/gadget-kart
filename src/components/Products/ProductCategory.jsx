import React, { useEffect } from "react";
import { connect } from "react-redux";
import Loading from "../../Loading";
import {
  fetchProductsByCategory,
  searchProduct,
} from "../../redux/products/products.action";
import SearchBar from "../SearchBar/SearchBar";
import CustomizeFields from "./CustomizeField";
import ProductList from "./ProductList";

export function ProductCategory({
  match: {
    params: { category },
  },
  fetchProductsByCategory,
  productsOriginal,
  products,
  fetchProductErr,
  searchProduct,
  productByCatLoading,
}) {
  useEffect(() => {
    fetchProductsByCategory(category);
  }, [fetchProductsByCategory, category]);

  const onSearch = (filterText) => {
    let filteredItems = productsOriginal.filter((item) =>
      item.title.toLocaleLowerCase().includes(filterText)
    );
    setTimeout(() => searchProduct(filteredItems), 500);
  };

  return (
    <div className="container my-4">
      {products && products.length > 0 ? (
        <>
          <h2 className="category-header">{category}</h2>
          <div className="row">
            <div className="col-10">
              <SearchBar onSearch={onSearch} />
            </div>
            <div style={{ position: "relative" }}>
              <button
                id="cutomizeBtn"
                className="btn btn-dark"
                style={{ style: "calc(1.5em + .75rem)" }}
                type="button"
                data-toggle="collapse"
                data-target="#customField"
                aria-controls="customField"
                aria-expanded="false"
                aria-label="Toggle navigation">
                <i className="fas fa-cog"></i>
                <span className="d-none d-md-inline-block ml-lg-3">
                  Customize
                </span>
              </button>
              <div className="collapse customize-collapse" id="customField">
                <CustomizeFields categoryName={category} />
              </div>
            </div>
          </div>

          <div className="row">
            <div className="col-12">
              <ProductList
                products={products}
                fields={{ ram: false, camera: false }}
              />
            </div>
          </div>
        </>
      ) : fetchProductErr ? (
        <h3>{fetchProductErr}</h3>
      ) : null}
      {productByCatLoading && (
        <div style={{ height: "60vh" }}>
          <Loading color="#ff4b2b" />
        </div>
      )}
    </div>
  );
}

const mapStateToProps = (state) => {
  return {
    productsOriginal: state.products.productsByCategoryOrig,
    products: state.products.productsByCategory,
    fetchProductErr: state.products.fetchProductErr,
    productByCatLoading: state.products.productByCatLoading,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    fetchProductsByCategory: (category) =>
      dispatch(fetchProductsByCategory(category)),
    searchProduct: (products) => dispatch(searchProduct(products)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ProductCategory);
