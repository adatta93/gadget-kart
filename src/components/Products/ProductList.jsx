import React from "react";
import Products from "./Products";
import "./Product.css";

function ProductList({ products, ...props }) {
  return (
    <div className="product-list-container">
      {products && products.length > 0 ? (
        <div className="product-list">
          {products.map((product) => (
            <Products key={product.id} product={product} {...props} />
          ))}
        </div>
      ) : null}
    </div>
  );
}

export default ProductList;
