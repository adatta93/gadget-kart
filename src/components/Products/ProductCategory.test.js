import { createBrowserHistory } from "history";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import Enzyme, { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";
import { Provider } from "react-redux";
import { applyMiddleware, createStore } from "redux";
import thunk from "redux-thunk";
import rootReducer from "../../redux/root.reducer";
import { ProductCategory } from "./ProductCategory";
import ProductList from "./ProductList";
import { Router } from "react-router-dom";

Enzyme.configure({ adapter: new Adapter() });

let store;
let props;
let history;
let products = [
  {
    id: 15,
    title: "Lenovo Ideapad S145 Core i3 7th Gen",
    img: "lenovo-ideapad.jpeg",
    category: "laptop",
    price: 31000,
    processor: "Intel Core i3 7th Gen",
    ram: "4GB",
    memory: "1TB HDD",
    screen: "15.6in Display",
    weight: "1.85kg",
  },
];
beforeAll(() => {
  props = {
    match: {
      params: {
        category: "Laptop",
      },
    },
    fetchProductsByCategory: jest.fn(),
    productsOriginal: products,
    products: products,
    fetchProductErr: "",
    searchProduct: jest.fn(),
    productByCatLoading: false,
  };
  store = createStore(rootReducer, applyMiddleware(thunk));
  history = createBrowserHistory();
});

describe("ProductCategory Snapshot", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ProductCategory {...props} />
        </Router>
      </Provider>
    );
  });

  it("renders correctly", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  afterEach(() => {
    wrapper.unmount();
  });
});

describe("ProductCategory rendering of elements", () => {
  let wrapper;
  beforeAll(() => {
    wrapper = shallow(<ProductCategory {...props} />);
  });

  it("renders category page header", () => {
    //console.log(wrapper.debug());
    expect(wrapper.find(".category-header").render().text()).toEqual("Laptop");
  });

  it("renders product list", () => {
    expect(wrapper.find(ProductList).length).toEqual(1);
  });

  it("renders customize field component on button click", () => {
    expect(wrapper.find("Connect(CustomizeFields)").length).toEqual(1);
  });
});
