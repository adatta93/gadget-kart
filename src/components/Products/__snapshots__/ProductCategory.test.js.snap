// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCategory Snapshot renders correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <ProductCategory
      fetchProductErr=""
      fetchProductsByCategory={
        [MockFunction] {
          "calls": Array [
            Array [
              "Laptop",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      match={
        Object {
          "params": Object {
            "category": "Laptop",
          },
        }
      }
      productByCatLoading={false}
      products={
        Array [
          Object {
            "category": "laptop",
            "id": 15,
            "img": "lenovo-ideapad.jpeg",
            "memory": "1TB HDD",
            "price": 31000,
            "processor": "Intel Core i3 7th Gen",
            "ram": "4GB",
            "screen": "15.6in Display",
            "title": "Lenovo Ideapad S145 Core i3 7th Gen",
            "weight": "1.85kg",
          },
        ]
      }
      productsOriginal={
        Array [
          Object {
            "category": "laptop",
            "id": 15,
            "img": "lenovo-ideapad.jpeg",
            "memory": "1TB HDD",
            "price": 31000,
            "processor": "Intel Core i3 7th Gen",
            "ram": "4GB",
            "screen": "15.6in Display",
            "title": "Lenovo Ideapad S145 Core i3 7th Gen",
            "weight": "1.85kg",
          },
        ]
      }
      searchProduct={[MockFunction]}
    >
      <div
        className="container my-4"
      >
        <h2
          className="category-header"
        >
          Laptop
        </h2>
        <div
          className="row"
        >
          <div
            className="col-10"
          >
            <SearchBar
              onSearch={[Function]}
            >
              <div
                className="input-group search-container mb-3"
              >
                <input
                  className="search-box"
                  onChange={[Function]}
                  placeholder="Search Product by Name"
                  type="search"
                  value=""
                />
                <div
                  className="input-group-append"
                >
                  <button
                    className="btn btn-search"
                    onClick={[Function]}
                    type="button"
                  >
                    <i
                      className="fas fa-search"
                    />
                  </button>
                </div>
              </div>
            </SearchBar>
          </div>
          <div
            style={
              Object {
                "position": "relative",
              }
            }
          >
            <button
              aria-controls="customField"
              aria-expanded="false"
              aria-label="Toggle navigation"
              className="btn btn-dark"
              data-target="#customField"
              data-toggle="collapse"
              id="cutomizeBtn"
              style={
                Object {
                  "style": "calc(1.5em + .75rem)",
                }
              }
              type="button"
            >
              <i
                className="fas fa-cog"
              />
              <span
                className="d-none d-md-inline-block ml-lg-3"
              >
                Customize
              </span>
            </button>
            <div
              className="collapse customize-collapse"
              id="customField"
            >
              <Connect(CustomizeFields)
                categoryName="Laptop"
              >
                <CustomizeFields
                  categoryName="Laptop"
                  fields={
                    Object {
                      "battery": true,
                      "camera": false,
                      "connector": true,
                      "memory": false,
                      "microphone": true,
                      "processor": false,
                      "ram": true,
                      "screen": false,
                      "sensor": true,
                      "speaker": true,
                      "weight": false,
                      "wireless": true,
                    }
                  }
                  updateCustomFields={[Function]}
                >
                  <div
                    className="card"
                  >
                    <div
                      className="card-header"
                    >
                      Customize Fields to Show
                    </div>
                    <div
                      className="card-body"
                    />
                  </div>
                </CustomizeFields>
              </Connect(CustomizeFields)>
            </div>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="col-12"
          >
            <ProductList
              fields={
                Object {
                  "camera": false,
                  "ram": false,
                }
              }
              products={
                Array [
                  Object {
                    "category": "laptop",
                    "id": 15,
                    "img": "lenovo-ideapad.jpeg",
                    "memory": "1TB HDD",
                    "price": 31000,
                    "processor": "Intel Core i3 7th Gen",
                    "ram": "4GB",
                    "screen": "15.6in Display",
                    "title": "Lenovo Ideapad S145 Core i3 7th Gen",
                    "weight": "1.85kg",
                  },
                ]
              }
            >
              <div
                className="product-list-container"
              >
                <div
                  className="product-list"
                >
                  <Connect(Products)
                    fields={
                      Object {
                        "camera": false,
                        "ram": false,
                      }
                    }
                    key="15"
                    product={
                      Object {
                        "category": "laptop",
                        "id": 15,
                        "img": "lenovo-ideapad.jpeg",
                        "memory": "1TB HDD",
                        "price": 31000,
                        "processor": "Intel Core i3 7th Gen",
                        "ram": "4GB",
                        "screen": "15.6in Display",
                        "title": "Lenovo Ideapad S145 Core i3 7th Gen",
                        "weight": "1.85kg",
                      }
                    }
                  >
                    <Products
                      dispatch={[Function]}
                      fields={
                        Object {
                          "battery": true,
                          "camera": false,
                          "connector": true,
                          "memory": false,
                          "microphone": true,
                          "processor": false,
                          "ram": true,
                          "screen": false,
                          "sensor": true,
                          "speaker": true,
                          "weight": false,
                          "wireless": true,
                        }
                      }
                      product={
                        Object {
                          "category": "laptop",
                          "id": 15,
                          "img": "lenovo-ideapad.jpeg",
                          "memory": "1TB HDD",
                          "price": 31000,
                          "processor": "Intel Core i3 7th Gen",
                          "ram": "4GB",
                          "screen": "15.6in Display",
                          "title": "Lenovo Ideapad S145 Core i3 7th Gen",
                          "weight": "1.85kg",
                        }
                      }
                    >
                      <Link
                        to={
                          Object {
                            "pathname": "/products/15",
                            "state": Object {
                              "prevPath": "/",
                            },
                          }
                        }
                      >
                        <LinkAnchor
                          href="/products/15"
                          navigate={[Function]}
                        >
                          <a
                            href="/products/15"
                            onClick={[Function]}
                          >
                            <div
                              className="card"
                            >
                              <div
                                className="card-body"
                              >
                                <div
                                  className="product-img-container"
                                >
                                  <img
                                    alt="Lenovo Ideapad S145 Core i3 7th Gen"
                                    className="card-img-top product-img"
                                    src="/images/lenovo-ideapad.jpeg"
                                  />
                                </div>
                                <h5
                                  className="card-title"
                                >
                                  Lenovo Ideapad S145 Core i3 7th Gen
                                </h5>
                                <h6
                                  className="card-subtitle"
                                >
                                  ₹31,000.00
                                </h6>
                                <div
                                  className="my-1"
                                  key="ram"
                                >
                                  <span
                                    className="prod-prop-name"
                                  >
                                    RAM
                                    :
                                  </span>
                                  <span
                                    className="prod-prop-value"
                                  >
                                    4GB
                                  </span>
                                </div>
                              </div>
                            </div>
                          </a>
                        </LinkAnchor>
                      </Link>
                    </Products>
                  </Connect(Products)>
                </div>
              </div>
            </ProductList>
          </div>
        </div>
      </div>
    </ProductCategory>
  </Router>
</Provider>
`;
